docs: revise documentation for best image practices (#11792)
sveltejs/kit: documentation/docs/40-best-practices/07-images.md

In some cases, the images may not be accessible at build time — e.g. they may live inside a content management system or elsewhere.

> Using a content delivery network (CDN) can allow you to optimize these images dynamically, and provides more flexibility with regards to sizes, but it may involve some setup overhead and usage costs. Depending on caching strategy, the browser may not be able to use a cached copy of the asset until a [304 response](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304) is received from the CDN. Building HTML to target CDNs may result in slightly smaller and simpler HTML because they can serve the appropriate file format for an `<img>` tag based on the `User-Agent` header whereas build-time optimizations must produce `<picture>` tags with multiple sources. However enabling brotli or gzip compression on the web server can mitigate the repetetive `<picture>` tag markup. Finally, some CDNs may generate images lazily, which could have a negative performance impact for sites with low traffic and frequently changing images.

Reviewer: There's a typo in "repetitive" here. I don't know if this sentence is worth adding though. I don't want to over-complicate things. The size is probably not all that important especially after compression as you point out. Maybe instead we can tweak the earlier sentence to remove the reference to the output being smaller
```suggestion
Using a content delivery network (CDN) can allow you to optimize these images dynamically, and provides more flexibility with regards to sizes, but it may involve some setup overhead and usage costs. Depending on caching strategy, the browser may not be able to use a cached copy of the asset until a [304 response](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304) is received from the CDN. Building HTML to target CDNs allows using an `<img>` tag since the CDN can serve the appropriate format based on the `User-Agent` header, whereas build-time optimizations must produce `<picture>` tags with multiple sources. Finally, some CDNs may generate images lazily, which could have a negative performance impact for sites with low traffic and frequently changing images.
```

- Consider serving all images via CDN regardless of the image optimization types you use. CDNs reduce latency by distributing copies of static assets globally.
- Your original images should have a good quality/resolution and should have 2x the width it will be displayed at to serve HiDPI devices. Image processing can size images down to save bandwidth when serving smaller screens, but it would be a waste of bandwidth to invent pixels to size images up.
- For images which are much larger than the width of a mobile device (roughly 400px), such as a hero image taking the width of the page design, specify `sizes` so that smaller images can be served on smaller devices.

> - Choose one image per page that is above the fold and the most important (or largest) and give it `fetchpriority="high" decoding="sync" loading="eager"` so it loads as early as possible. This gives you better web vitals scores (largest contentful paint, in particular).

Reviewer: I hadn't seen the `decoding="sync"` suggestion before. Any references you can share about that?
```suggestion
- Choose one image per page that is above the fold and the most important (or largest) and give it `fetchpriority="high" decoding="sync" loading="eager"` so it loads as early as possible. This gives you better web vitals scores (largest contentful paint (LCP), in particular).
```

> - Give the image a container or styling so that it is constrained and does not jump around while the page is loading causing a cumulative layout shift. `width` and `height` help the browser to reserve a "slot" of space as it uses these values to determine an aspect ratio before the CSS object model has had a chance to be created and while the image is still loading. `@sveltejs/enhanced-img` will add a `width` and `height` for you for this reason.

Reviewer: this seems overly complicated. I think we just need to tell people what to do - not educate them about the browser internals
```suggestion
- Give the image a container or styling so that it is constrained and does not jump around while the page is loading affecting your cumulative layout shift metric. `width` and `height` help the browser to reserve space while the image is still loading, so `@sveltejs/enhanced-img` will add a `width` and `height` for you.
```

- Always provide a good `alt` text. The Svelte compiler will warn you if you don't do this.

> - When used in `sizes` or `@media` queries, `em` and `rem` are both specced to mean the user's default `font-size`. For a `sizes` decalaration like `sizes="(min-width: 768px) min(100vw, 108rem), 64rem"`, the actual `em` or `rem` that controls how the image is laid out on the page can end up different if your CSS changes it. This means one should not change the default size of `em` if they want to give the image preloader truthful information. For example, do not do something like `html { font-size: 62.5%; }` as the slot reserved by the browser preloader will now end up being larger than the actual slot of the CSS object model once it has been created.

Reviewer: the first sentence should be the recommendation. the explanation can follow
```suggestion
- Do not use `em` or `rem` in `sizes` and change the default size of these measures. When used in `sizes` or `@media` queries, `em` and `rem` are both defined to mean the user's default `font-size`. For a `sizes` declaration like `sizes="(min-width: 768px) min(100vw, 108rem), 64rem"`, the actual `em` or `rem` that controls how the image is laid out on the page can be different if changed by CSS. For example, do not do something like `html { font-size: 62.5%; }` as the slot reserved by the browser preloader will now end up being larger than the actual slot of the CSS object model once it has been created.
```
